File: driver/runner.js
/**
 * Driver / Runner – Central Execution Controller
 * Mirrors the Driver/Runner (ExecutionEngine) from the Jarvis framework.
 *
 * Responsibilities:
 *   - Initialize / teardown LeanFT SDK
 *   - Manage the TE session lifecycle
 *   - Provide a shared session handle to all test specs
 *   - Apply execution settings (timeouts, screenshots on failure)
 *
 * Jasmine test specs use this runner via beforeAll / afterAll hooks
 * rather than managing SDK lifecycle themselves.
 *
 * Usage (in a spec file):
 *   const runner = require("../driver/runner");
 *   beforeAll(runner.setup);
 *   afterAll(runner.teardown);
 *   it("...", async () => { const session = runner.getSession(); });
 */

"use strict";

const LFT      = require("leanft");
const settings = require("../config/settings");
const terminal = require("../libraries/terminalHelper");
const reporter = require("../libraries/reportHelper");

let _session = null;

/**
 * Initialize LeanFT and open a mainframe TE session.
 * Bind to Jasmine beforeAll.
 */
async function setup() {
    // Init LeanFT SDK
    await LFT.init({
        host: settings.agent.host,
        port: settings.agent.port
    });

    reporter.log("LeanFT SDK initialized");
    reporter.log(`Environment: ${settings.environment}`);

    // Open mainframe session
    _session = await terminal.openSession();
}

/**
 * Sign off, close session, and cleanup LeanFT.
 * Bind to Jasmine afterAll.
 */
async function teardown() {
    try {
        if (_session) {
            await terminal.closeSession(_session);
        }
    } catch (e) {
        reporter.log(`Teardown warning: ${e.message}`);
    } finally {
        await LFT.cleanup();
        reporter.log("LeanFT SDK cleaned up");
    }
}

/**
 * Hook to run before each test.
 * Captures screenshot on failure if configured.
 */
async function beforeEachTest() {
    await LFT.beforeTest();
}

/**
 * Hook to run after each test.
 * Captures screenshot on failure if configured.
 */
async function afterEachTest() {
    await LFT.afterTest();
}

/**
 * Returns the active TE session for use in test steps.
 * @returns {TeSession}
 */
function getSession() {
    if (!_session) {
        throw new Error("No active TE session. Ensure runner.setup() completed successfully.");
    }
    return _session;
}

module.exports = {
    setup,
    teardown,
    beforeEachTest,
    afterEachTest,
    getSession
};

File: driver/runner.py
"""
Driver / Runner
Mirrors the ExecutionEngine / Driver/Runner layer from the Jarvis framework.

Manages the TN3270 session lifecycle for the entire pytest session:
  - session_setup()    → open TN3270 connection (runs once before all tests)
  - session_teardown() → close TN3270 connection (runs once after all tests)
  - before_each()      → log test start, capture initial screen state
  - after_each()       → log result, capture screen on failure

Used via the conftest.py pytest fixtures — tests never call this directly.
"""

from py3270 import Emulator
from libraries import terminal_helper as terminal
from libraries import report_helper as reporter
from config import settings

# Module-level reference to the active emulator (shared across all tests)
_emulator: Emulator | None = None


def get_emulator() -> Emulator:
    """Return the active Emulator instance. Raises if session is not open."""
    if _emulator is None:
        raise RuntimeError(
            "TN3270 session is not open. "
            "Ensure session_setup() has been called (check conftest.py fixtures)."
        )
    return _emulator


def session_setup() -> Emulator:
    """
    Open TN3270 session and store reference.
    Called once before the full test suite (session-scoped fixture).
    """
    global _emulator
    reporter.log("=== SESSION SETUP ===", f"ENV: {settings.TEST_ENV}  Host: {settings.HOST}:{settings.PORT}")
    _emulator = terminal.open_session()
    return _emulator


def session_teardown() -> None:
    """
    Close TN3270 session and write the HTML report.
    Called once after all tests complete (session-scoped fixture).
    """
    global _emulator
    reporter.log("=== SESSION TEARDOWN ===")
    terminal.close_session(_emulator)
    _emulator = None
    reporter.write_html_report()


def before_each_test(test_name: str) -> None:
    """Log test start. Called from pytest fixture before each test."""
    reporter.log(f"--- Before: {test_name} ---")


def after_each_test(test_name: str, passed: bool) -> None:
    """Capture screenshot on failure, log result. Called from pytest fixture."""
    em = _emulator
    if not passed and em and settings.SCREENSHOT_ON_FAILURE:
        try:
            terminal.capture_screen(em, f"FAILURE screenshot — {test_name}")
        except Exception:
            pass
    status = "PASS" if passed else "FAIL"
    reporter.log(f"--- After: {test_name} — {status} ---")
